README.md
<div>
  <div align="center" style="display: block; text-align: center;">
    <img src="https://camo.githubusercontent.com/734a3468bce992fbc3b729562d41c92f4912c99a/68747470733a2f2f7777772e727573742d6c616e672e6f72672f7374617469632f696d616765732f727573742d6c6f676f2d626c6b2e737667" height="120" width="120" />
  </div>
  <h1 align="center">rifven</h1>
  <h4 align="center">
    Venezuelan RIF implementation useful for creating and validating RIF numbers
  </h4>
</div>

<div align="center">

  [![Crates.io](https://img.shields.io/crates/v/rifven.svg)](https://crates.io/crates/rifven)
  [![Documentation](https://docs.rs/rifven/badge.svg)](https://docs.rs/rifven)
  ![Build](https://github.com/EstebanBorai/rifven/workflows/build/badge.svg)
  ![Lint](https://github.com/EstebanBorai/rifven/workflows/clippy/fmt/badge.svg)
  ![Tests](https://github.com/EstebanBorai/rifven/workflows/tests/badge.svg)

</div>

## What are RIF numbers?

RIF (Registro de Informacion Fiscal) in english _Fiscal Information Registry_ is a number
provided by a Venezuelan entity SAIME used to identify multiple entities for taxable purposes.

The RIF is composed by a kind which could be:

- C: Township or Communal Council
- E: Represents a foreigner natural person and stands for
"Extranjera" and "Extranjero"
- G: Represents a goverment entity and stands for
"Gubernamental"
- J: Used for a legal entity. Could be a natural person
or a corporate entity and stands for "Jurídico"
- P: Used on RIF numbers which belongs to passports
- V: Represents a person with venezuelan citizenship and stands
for "Venezolana" and "Venezolano"

An identifier number followed by a hyphen symbol and finally a checksum digit, as well followed
by a hyphen symbol.

<div align="center">
  <img src="https://raw.githubusercontent.com/EstebanBorai/rifven/main/docs/rif_parts.png" />
</div>

## Motivation

Implement a crate to help create instances of valid RIF numbers

## Usage

Creating a new `Rif` instance providing each of its parts values
such as `Kind` (J; V; P; G; C), identifier (tax payer ID), check number.

The following code, creates an instance of `Rif` for a RIF string which
looks like `J-07013380-5`:

```rust
use rifven::{Kind, Rif};

let rif = Rif::new(Kind::Legal, 07013380, 5).unwrap();

assert_eq!(rif.kind(), Kind::Legal);
assert_eq!(rif.identifier(), 7013380);
assert_eq!(rif.checksum_digit(), 5);
```

You can also create instances of `Rif` from its string representation

```rust
use rifven::{Kind, Rif};
use std::str::FromStr;

let myrif = Rif::from_str("J-07013380-5").unwrap();

assert_eq!(Rif::new(Kind::Legal, 07013380, 5).unwrap(), myrif);
```

## License

Licensed under the Apache 2.0 and MIT licenses

## Contributions

Every contribution to this project is welcome, feel free to either open a Pull Request
or issue


CHANGELOG.md
## v0.1.0

- Add `Rif` `struct` with validation
- Add `FromStr` implementation for `Rif`
- Add `Kind` `enum` with support for foreigner, government,
legal entity, passport, township and venezuelan types
- Add `FromStr` implementation for `Kind`
- Add capability to calculate the _check number_

src/rif.rs
use std::ops::IndexMut;
use std::str::FromStr;

use crate::error::{Error, Result};
use crate::kind::Kind;

/// Length of a RIF identifier
const RIF_IDENTIFIER_LENGTH: usize = 9;

/// Venezuelan RIF identifier
///
/// # Anatomy
/// The following chart reviews the anatomy of a
/// RIF identifier.
///
///```ignore
///
/// X - XXXXXXXXX - X
/// ^   ^________   ^
/// A       B      C
///
/// ```
///
/// A: `Kind` represents either the citizenship or the
/// kind of entity behind the RIF identifier
///
/// B: The contributor/tax payer identifier. Must be 9
/// digits long
///
/// C: The check number for the RIF identifier
#[derive(Clone, Debug, PartialEq)]
pub struct Rif {
    pub(crate) checksum_digit: u8,
    pub(crate) identifier: u32,
    pub(crate) kind: Kind,
}

impl Rif {
    /// Creates a new `Rif` instance with the provided `kind`, `identifier`
    /// and the `checksum_digit`
    ///
    /// ```rust
    /// use rifven::{Kind, Rif};
    ///
    /// let rif = Rif::new(Kind::Legal, 07013380, 5).unwrap();
    ///
    /// assert_eq!(rif.kind(), Kind::Legal);
    /// assert_eq!(rif.identifier(), 7013380);
    /// assert_eq!(rif.checksum_digit(), 5);
    /// ```
    pub fn new(kind: Kind, identifier: u32, checksum_digit: u8) -> Result<Self> {
        let expected_checksum_digit = Rif::calc_checksum_digit(&kind, identifier);

        if expected_checksum_digit == checksum_digit {
            return Ok(Rif {
                checksum_digit,
                identifier,
                kind,
            });
        }

        Err(Error::UnexpectedCheckNum(
            expected_checksum_digit,
            checksum_digit,
        ))
    }

    /// Retrieves the `Kind` of the current `Rif`
    ///
    /// ```rust
    /// use rifven::{Kind, Rif};
    ///
    /// let rif = Rif::new(Kind::Legal, 07013380, 5).unwrap();
    ///
    /// assert_eq!(rif.kind(), Kind::Legal);
    /// ```
    pub fn kind(&self) -> Kind {
        self.kind.clone()
    }

    /// Retrieves the identifier of the current `Rif`
    ///
    /// ```rust
    /// use rifven::{Kind, Rif};
    ///
    /// let rif = Rif::new(Kind::Legal, 07013380, 5).unwrap();
    ///
    /// assert_eq!(rif.identifier(), 7013380);
    /// ```
    pub fn identifier(&self) -> u32 {
        self.identifier
    }

    /// Retrieves the check number of the current `Rif`
    ///
    /// ```rust
    /// use rifven::{Kind, Rif};
    ///
    /// let rif = Rif::new(Kind::Legal, 07013380, 5).unwrap();
    ///
    /// assert_eq!(rif.checksum_digit(), 5);
    /// ```
    pub fn checksum_digit(&self) -> u8 {
        self.checksum_digit
    }

    /// Validates the RIF provided as a string slice (`&str`).
    /// Returning a `Result` where the `Ok` variant contains a
    /// `Rif` instance built from the `rif` string slice and the
    /// `Err` variant contains the `Error` (`rifven::error::Error`)
    /// instance created during the validation
    ///
    /// The RIF provided must be separated using a hyphen (`-`) as
    /// follows `J-40512535-7` otherwise this function will return
    /// an `Err` variant containing the `Error::InvalidRif` error
    fn validate(rif: &str) -> Result<Rif> {
        let parts: Vec<&str> = rif.split("-").collect();

        if parts.len() != 3 {
            return Err(Error::InvalidRif(format!("RIF must be splitted into 3 parts separated by dashes. Eg. J-123456789-1. Provided {}", rif)));
        }

        let checksum_digit = parts
            .get(2)
            .unwrap()
            .parse::<u8>()
            .map_err(|_| Error::InvalidCheckNum(parts.get(2).unwrap().to_string()))?;
        let kind = Kind::from_str(parts.get(0).unwrap())?;
        let identifier = {
            let identifier = parts.get(1).unwrap();

            identifier
                .parse::<u32>()
                .map_err(|e| Error::InvalidRifIdentifier(e.to_string()))?
        };
        let expected_checksum_digit = Rif::calc_checksum_digit(&kind, identifier);

        if expected_checksum_digit != checksum_digit {
            return Err(Error::UnexpectedCheckNum(
                expected_checksum_digit,
                checksum_digit,
            ));
        }

        Ok(Rif {
            checksum_digit,
            identifier,
            kind,
        })
    }

    /// Calculates the **checksum_digit** for a provided RIF number
    pub fn calc_checksum_digit(kind: &Kind, identifier: u32) -> u8 {
        let mut digits: Vec<u32> = vec![0; RIF_IDENTIFIER_LENGTH];
        let mut sum_values: Vec<u32> = vec![0; RIF_IDENTIFIER_LENGTH];
        let mut identifier = identifier;

        for idx in 1..=RIF_IDENTIFIER_LENGTH {
            *digits.index_mut(RIF_IDENTIFIER_LENGTH - idx) = identifier % 10;
            identifier /= 10;
        }

        for (idx, digit) in digits.into_iter().enumerate() {
            let current_value = sum_values.index_mut(idx);

            match idx {
                0 => *current_value = kind.checksum_digit() * 4,
                1 | 7 => *current_value = digit * 3,
                2 | 8 => *current_value = digit * 2,
                3 => *current_value = digit * 7,
                4 => *current_value = digit * 6,
                5 => *current_value = digit * 5,
                6 => *current_value = digit * 4,
                _ => {}
            }
        }

        let sum_values_total: u32 = sum_values.iter().sum();
        let validator = sum_values_total / 11;
        let reminder = sum_values_total - (validator * 11);
        let checksum_digit = 11 - reminder;

        if checksum_digit > 9 {
            return 0;
        }

        checksum_digit as u8
    }
}

impl FromStr for Rif {
    type Err = Error;

    fn from_str(s: &str) -> std::result::Result<Self, Self::Err> {
        Ok(Rif::validate(s)?)
    }
}


src/kind.rs
use std::str::FromStr;
use std::string::ToString;

use crate::error::Error;

/// The RIF kind represents the kind of entity this RIF
/// belongs to.
///
/// Possible values are:
///
/// - C: Township or Communal Council
/// - E: Represents a foreigner natural person and stands for
/// "Extranjera" and "Extranjero"
/// - G: Represents a goverment entity and stands for
/// "Gubernamental"
/// - J: Used for a legal entity. Could be a natural person
/// or a corporate entity and stands for "Jurídico"
/// - P: Used on RIF numbers which belongs to passports
/// - V: Represents a person with venezuelan citizenship and stands
/// for "Venezolana" and "Venezolano"
#[derive(Clone, Debug, PartialEq)]
pub enum Kind {
    /// E: Foreigner
    Foreigner,
    /// G: Government Entity
    Government,
    /// J: Legal Entity
    Legal,
    /// P: Passport
    Passport,
    /// C: Township or Communal Council
    Township,
    /// V: Venezuelan
    Venezuelan,
}

impl Kind {
    /// Retrieves the checksum digit for the current RIF kind
    pub fn checksum_digit(&self) -> u32 {
        match self {
            Kind::Venezuelan => 1,
            Kind::Foreigner => 2,
            Kind::Legal | Kind::Township => 3,
            Kind::Passport => 4,
            Kind::Government => 5,
        }
    }
}

impl FromStr for Kind {
    type Err = Error;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s.to_lowercase().as_str() {
            "c" => Ok(Kind::Township),
            "e" => Ok(Kind::Foreigner),
            "g" => Ok(Kind::Government),
            "j" => Ok(Kind::Legal),
            "p" => Ok(Kind::Passport),
            "v" => Ok(Kind::Venezuelan),
            _ => Err(Error::InvalidRifKind(s.to_string())),
        }
    }
}

impl ToString for Kind {
    fn to_string(&self) -> String {
        match self {
            Kind::Foreigner => String::from("E"),
            Kind::Government => String::from("G"),
            Kind::Legal => String::from("J"),
            Kind::Passport => String::from("P"),
            Kind::Township => String::from("C"),
            Kind::Venezuelan => String::from("V"),
        }
    }
}

#[allow(unused_imports)]
mod tests {
    use super::*;

    #[test]
    fn creates_a_kind_instance_from_str() {
        let kinds = vec![
            Kind::Foreigner,
            Kind::Government,
            Kind::Legal,
            Kind::Passport,
            Kind::Township,
            Kind::Venezuelan,
        ];

        let str_value = vec!["e", "g", "j", "p", "c", "v"];

        for (idx, kind_str) in str_value.into_iter().enumerate() {
            assert_eq!(kinds[idx], Kind::from_str(kind_str).unwrap());
        }
    }

    #[test]
    fn creates_a_string_from_a_kind() {
        let kinds = vec![
            Kind::Township,
            Kind::Foreigner,
            Kind::Government,
            Kind::Legal,
            Kind::Passport,
            Kind::Venezuelan,
        ];

        let string_value = vec![
            String::from("C"),
            String::from("E"),
            String::from("G"),
            String::from("J"),
            String::from("P"),
            String::from("V"),
        ];

        for (idx, kind) in kinds.into_iter().enumerate() {
            assert_eq!(string_value[idx].to_string(), kind.to_string());
        }
    }
}


src/lib.rs
//! # rifve
//!
//! Venezuelan RIF implementation useful for creating and validating RIF numbers
//!
//! <div align="center">
//! 
//!   [![Crates.io](https://img.shields.io/crates/v/rifven.svg)](https://crates.io/crates/rifven)
//!   [![Documentation](https://docs.rs/rifven/badge.svg)](https://docs.rs/rifven)
//!   ![Build](https://github.com/EstebanBorai/rifven/workflows/build/badge.svg)
//!   ![Lint](https://github.com/EstebanBorai/rifven/workflows/clippy/fmt/badge.svg)
//!   ![Tests](https://github.com/EstebanBorai/rifven/workflows/tests/badge.svg)
//! 
//! </div>
//! 
//! ## What are RIF numbers?
//! 
//! RIF (Registro de Informacion Fiscal) in english _Fiscal Information Registry_ is a number
//! provided by a Venezuelan entity SAIME used to identify multiple entities for taxable purposes.
//! 
//! The RIF is composed by a kind which could be:
//! 
//! - C: Township or Communal Council
//! - E: Represents a foreigner natural person and stands for
//! "Extranjera" and "Extranjero"
//! - G: Represents a goverment entity and stands for
//! "Gubernamental"
//! - J: Used for a legal entity. Could be a natural person
//! or a corporate entity and stands for "Jurídico"
//! - P: Used on RIF numbers which belongs to passports
//! - V: Represents a person with venezuelan citizenship and stands
//! for "Venezolana" and "Venezolano"
//! 
//! An identifier number followed by a hyphen symbol and finally a checksum digit, as well followed
//! by a hyphen symbol.
//! 
//! <div align="center">
//!   <img src="https://raw.githubusercontent.com/EstebanBorai/rifven/main/docs/rif_parts.png" />
//! </div>
//! 
//! ## Motivation
//! 
//! Implement a crate to help create instances of valid RIF numbers
//! 
//! ## Usage
//! 
//! Creating a new `Rif` instance providing each of its parts values
//! such as `Kind` (J; V; P; G; C), identifier (tax payer ID), check number.
//!
//! The following code, creates an instance of `Rif` for a RIF string which
//! looks like `J-07013380-5`:
//!
//! ```
//! use rifven::{Kind, Rif};
//!
//! let rif = Rif::new(Kind::Legal, 07013380, 5).unwrap();
//!
//! assert_eq!(rif.kind(), Kind::Legal);
//! assert_eq!(rif.identifier(), 7013380);
//! assert_eq!(rif.checksum_digit(), 5);
//! ```
//!
//! You can also create instances of `Rif` from its string representation
//!
//! ```
//! use rifven::{Kind, Rif};
//! use std::str::FromStr;
//!
//! let myrif = Rif::from_str("J-07013380-5").unwrap();
//!
//! assert_eq!(Rif::new(Kind::Legal, 07013380, 5).unwrap(), myrif);
//! ```
//!
mod error;
mod kind;
mod rif;

pub use kind::*;
pub use rif::*;

#[allow(unused_imports)]
mod tests {
    use std::str::FromStr;

    use crate::error::*;
    use crate::kind::*;
    use crate::rif::*;

    #[test]
    fn calcs_the_checksum_digit() {
        let rif_identifier = vec![
            (Kind::Legal, 00019361),
            (Kind::Legal, 07013380),
            (Kind::Legal, 31286704),
            (Kind::Legal, 40512535),
            (Kind::Legal, 40119253),
            (Kind::Government, 20009997),
            (Kind::Government, 20000001),
            (Kind::Government, 20000002),
            (Kind::Government, 20000004),
            (Kind::Government, 20000044),
        ];

        let rif_checksum_digit = vec![4, 5, 3, 7, 0, 6, 5, 3, 0, 9];

        for (idx, (kind, identifier)) in rif_identifier.iter().enumerate() {
            assert_eq!(
                rif_checksum_digit[idx],
                Rif::calc_checksum_digit(kind, *identifier),
            );
        }
    }

    #[test]
    fn creates_rif_from_str() {
        let candidates = vec![
            Rif::new(Kind::Legal, 000019361, 4).unwrap(),
            Rif::new(Kind::Legal, 07013380, 5).unwrap(),
            Rif::new(Kind::Legal, 31286704, 3).unwrap(),
            Rif::new(Kind::Government, 20000044, 9).unwrap(),
            Rif::new(Kind::Government, 20000004, 0).unwrap(),
            Rif::new(Kind::Government, 20000002, 3).unwrap(),
        ];

        let expects = vec![
            Rif::from_str("J-00019361-4").unwrap(),
            Rif::from_str("J-07013380-5").unwrap(),
            Rif::from_str("J-31286704-3").unwrap(),
            Rif::from_str("G-20000044-9").unwrap(),
            Rif::from_str("G-20000004-0").unwrap(),
            Rif::from_str("G-20000002-3").unwrap(),
        ];

        for (idx, rif) in candidates.iter().enumerate() {
            assert_eq!(*rif, expects[idx]);
        }
    }

    #[test]
    fn checks_for_invalid_rifs_from_str() {
        let have = vec![
            Rif::from_str("J-00018461-4"),
            Rif::from_str("E-12312312-5"),
            Rif::from_str("M-00000001-3"),
            Rif::from_str("X-00029383-7"),
            Rif::from_str("V-AA348932-1"),
            Rif::from_str("G-X0000002-3"),
            Rif::from_str("G200000040"),
        ];

        let expected_error = vec![
          Error::UnexpectedCheckNum(5, 4),
          Error::UnexpectedCheckNum(6, 5),
          Error::InvalidRifKind(String::from("M")),
          Error::InvalidRifKind(String::from("X")),
          Error::InvalidRifIdentifier(String::from("invalid digit found in string")),
          Error::InvalidRifIdentifier(String::from("invalid digit found in string")),
          Error::InvalidRif(String::from("RIF must be splitted into 3 parts separated by dashes. Eg. J-123456789-1. Provided G200000040")),
      ];

        for (idx, rif) in have.into_iter().enumerate() {
            assert_eq!(rif.err().unwrap(), expected_error[idx]);
        }
    }
}
